### PR TITLE
Tests for wp_viewporter

### PR DESCRIPTION
**Pull request by RAOF**

### CODE REVIEW

File: tests/wp_viewporter.cpp [hunk 1]
```diff
@@ -0,0 +1,553 @@
...
+    auto surface = client.create_visible_surface(buffer_width, buffer_height);
+    auto buffer = ShmBuffer(client, buffer_width, buffer_height);
+
+    auto viewporter = client.bind_if_supported<wp_viewporter>(wlcs::AnyVersion);
+    auto viewport = wp_viewporter_get_viewport(viewporter, surface);
+
+    bool committed = false;
+
+    wp_viewport_set_destination(viewport, display_width, display_height);
+    // Oh! I think we handle changing surface properties incorrectly!
```
Comment (mattkae) [inline tests/wp_viewporter.cpp:145]:
```suggestion
```

File: tests/wp_viewporter.cpp [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+    client.dispatch_until([&committed]() { return committed;} );
+
+    EXPECT_TRUE(surface_has_size(client, surface, display_width, display_height));
+}
+
+TEST_F(WpViewporterTest, when_buffer_is_scaled_destination_is_in_scaled_coordinates)
+{
+    wlcs::Client client{the_server()};
+
+
```
Comment (mattkae) [inline tests/wp_viewporter.cpp:206]:
```suggestion
```

File: tests/wp_viewporter.cpp [hunk 3]
```diff
@@ -0,0 +1,553 @@
...
+}
+
+#define EXPECT_PROTOCOL_ERROR(block, iface, err_code) \
+    try { \
+        block \
+        FAIL() << "Expected protocol error not raised"; \
+    } catch (wlcs::ProtocolError const& err) {\
+        EXPECT_THAT(err.interface(), Eq(iface));\
+        EXPECT_THAT(err.error_code(), Eq(err_code));\
+    }
```
Comment (mattkae) [inline tests/wp_viewporter.cpp:268]: Would this be a good general WLCS utility for other tests?

File: tests/wp_viewporter.cpp [hunk 4]
```diff
@@ -0,0 +1,553 @@
...
+        {
+            wp_viewport_set_destination(viewport, 10, 10);
+            client.roundtrip();
+        },
+        &wp_viewport_interface,
+        WP_VIEWPORT_ERROR_NO_SURFACE
+    );
+}
+
+TEST_F(WpViewporterTest, setting_surface_after_surface_has_been_destroyed_is_an_error)
```
Comment (mattkae) [inline tests/wp_viewporter.cpp:364]:
```suggestion
TEST_F(WpViewporterTest, setting_source_after_surface_has_been_destroyed_is_an_error)
```

**Review by mattkae — CHANGES_REQUESTED** [review]
Some minor things, but looks good!

File: tests/wp_viewporter.cpp (hunk 3 above)
Comment (RAOF) [reply tests/wp_viewporter.cpp]: It probably would! Now in its own header.

**Review by mattkae — APPROVED** [review]
Looks good to me too!

**Pull request merged**